File: src/main/java/org/retrolauncher/backend/app/games/infrastructure/services/PSXGameFinderService.java
package org.retrolauncher.backend.app.games.infrastructure.services;

import org.retrolauncher.backend.app.games.application.services.GameFinderService;

import java.io.*;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Pattern;

public class PSXGameFinderService implements GameFinderService {
    @Override
    public List<File> getFilesFrom(File folder) {
        final var files = folder.listFiles();
        final var identifiedFiles = new ArrayList<File>();

        if (files == null)
            return new ArrayList<>();

        for (var file : files) {
            if (!isCUEFile(file)) continue;

            final var binFile = getBinFileFrom(file);

            if (binFile.isEmpty()) continue;
            if (!isPSXFile(binFile.get())) continue;

            identifiedFiles.add(file);
        }

        return identifiedFiles;
    }

    private boolean isCUEFile(File file) {
        return file.getName().contains(".cue");
    }

    private Optional<File> getBinFileFrom(File cueFile) {
        final var pattern = Pattern.compile("FILE \"([^\"]+)\" BINARY");
        final var binFileName = searchForContentOnFile(cueFile, pattern);

        if (binFileName.isEmpty()) return Optional.empty();

        final var binFile = Path.of(cueFile.getParent()).resolve(binFileName.get()).toFile();

        return Optional.ofNullable(binFile.exists() ? binFile : null);
    }

    private boolean isPSXFile(File binFile) {
        return searchForContentOnFile(binFile, Pattern.compile("(1993-1997 Sony)")).isPresent();
    }

    private Optional<String> searchForContentOnFile(File file, Pattern pattern) {
        String line;
        FileInputStream inputStream = null;
        String result = null;

        try {
            inputStream = new FileInputStream(file);
            final BufferedReader buffer = new BufferedReader(new InputStreamReader(inputStream));

            while ((line = buffer.readLine()) != null) {
                final var matcher = pattern.matcher(line);

                if (matcher.find()) {
                    result = matcher.group(1);
                    break;
                }
            }
        } catch (IOException exception) {
            Logger.getLogger(this.getClass().getName()).log(Level.WARNING, exception.getMessage());
        }

        closeInputStream(inputStream);

        return Optional.ofNullable(result);
    }

    private void closeInputStream(InputStream inputStream) {
        if (inputStream == null)
            return;

        try {
            inputStream.close();
        } catch (IOException exception) {
            Logger.getLogger(this.getClass().getName()).log(Level.WARNING, exception.getMessage());
        }
    }
}

File: backend/src/app/games/internal/services/psx_game_finder_service.go
package services

import (
	"bufio"
	"os"
	"path/filepath"
	"regexp"
	"retrolauncher/backend/src/app/games/internal/application"
	"strings"
)

type psxGameFinderService struct{}

func NewPSXGameFinderService() application.GameFinderService {
	return &psxGameFinderService{}
}

func (s *psxGameFinderService) GetFilesFrom(folder string) []string {
	files, err := os.ReadDir(folder)
	if err != nil {
		return []string{}
	}

	var identifiedFiles []string
	for _, file := range files {
		if file.IsDir() {
			continue
		}

		filePath := filepath.Join(folder, file.Name())

		if !s.isCUEFile(file.Name()) {
			continue
		}

		binFile, ok := s.getBinFileFrom(filePath)
		if !ok {
			continue
		}

		if !s.isPSXFile(binFile) {
			continue
		}

		identifiedFiles = append(identifiedFiles, filePath)
	}

	return identifiedFiles
}

func (s *psxGameFinderService) isCUEFile(fileName string) bool {
	return strings.HasSuffix(strings.ToLower(fileName), ".cue")
}

func (s *psxGameFinderService) getBinFileFrom(cueFilePath string) (string, bool) {
	pattern := regexp.MustCompile(`FILE "([^"]+)" BINARY`)
	binFileName := s.searchForContentOnFile(cueFilePath, pattern)

	if binFileName == "" {
		return "", false
	}

	binFilePath := filepath.Join(filepath.Dir(cueFilePath), binFileName)

	if _, err := os.Stat(binFilePath); os.IsNotExist(err) {
		return "", false
	}

	return binFilePath, true
}

func (s *psxGameFinderService) isPSXFile(binFilePath string) bool {
	pattern := regexp.MustCompile(`(Sony Computer Entertainment)`)
	return s.searchForContentOnFile(binFilePath, pattern) != ""
}

func (s *psxGameFinderService) searchForContentOnFile(filePath string, pattern *regexp.Regexp) string {
	file, err := os.Open(filePath)
	if err != nil {
		return ""
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	// Aumentar o buffer para arquivos binários grandes
	buf := make([]byte, 0, 64*1024)
	scanner.Buffer(buf, 1024*1024)

	for scanner.Scan() {
		line := scanner.Text()
		matches := pattern.FindStringSubmatch(line)
		if len(matches) > 1 {
			return matches[1]
		}
	}

	return ""
}
